### PR TITLE
Don't lint `manual_non_exhaustive` when the enum variant is used

### LINKED ISSUE #5714: manual_non_exhaustive should not warn on a used enum variant
I have an enum similar to the following:

```rust

enum State {
        /// Temporary value to use in place of uninitialized memory, for safe handling of panics
        #[doc(hidden)]
        _Ephemeral,

        Uninit(...),
        Init(...),
    }
```

Clippy suggests that I remove the `_Ephemeral` variant in place of `#[non_exhaustive]`, which wouldn't work in this case as `_Ephemeral` is used, so removing it would be an error. I think the lint should only complain if the variant isn't actually used?

```
cargo clippy -V
clippy 0.0.212 (feb3536 2020-06-09)
```
Thanks!

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #5714

changelog: Don't lint `manual_non_exhaustive` when the enum variant is used

### CODE REVIEW

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
This looks good to me, thank you for the implementation!

I've marked one question, if this change was intentional, you can `r=me` (`@bors r=xFrednet`) this PR :upside_down_face: 

@bors d+

**Comment by xFrednet** [comment]
Looks like I messed up with some commands. Let's try this again:

@bors delegate+

File: clippy_lints/src/manual_non_exhaustive.rs [hunk 1]
```diff
@@ -58,129 +62,159 @@ declare_clippy_lint! {
...
+            }
+            let mut iter = fields.iter().filter_map(|f| match f.vis.kind {
+                VisibilityKind::Public => None,
+                VisibilityKind::Inherited => Some(Ok(f)),
+                _ => Some(Err(())),
+            });
+            if let Some(Ok(field)) = iter.next()
+                && iter.next().is_none()
+                && field.ty.kind.is_unit()
+                && field.ident.map_or(true, |name| name.as_str().starts_with('_'))
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/manual_non_exhaustive.rs:122]: The test for `#[doc(hidden)]` didn't exist for structs, only enums.

**Pull request merged**